chrome/content/zotero/components/tagSelector/tagSelectorList.jsx

```diff
@@ -214,6 +221,70 @@ class TagList extends React.PureComponent {
 			</div>
 		);
 	};
+
+	shiftFocus(e) {
+		if (!document.activeElement.classList.contains("tag-selector-item")) return;
+		if (!["ArrowRight", "ArrowLeft"].includes(e.key)) return;
+		// Handle arrow navigation
+		let nextTag = (node) => {
+			if (e.key == "ArrowRight") return node.nextElementSibling;
+			return node.previousElementSibling;
+		};
+		let nextOne = nextTag(document.activeElement);
+		// Skip disabled tags
+		while (nextOne && nextOne.classList.contains("disabled")) {
+			nextOne = nextTag(nextOne);
+		}
+		if (nextOne) {
+			nextOne.focus();
+		}
+	}
+
+	handleSectionRendered = ({ indices }) => {
+		let tagsList = document.querySelector('.tag-selector-list');
+		// <Collection> sets role="grid" which is not semantically correct
+		tagsList.setAttribute("role", "group");
+
+		this.setState({ scrollToCell: undefined });
+		if (this.focusedTagIndex === null) return;
+		// Check if the tag that is supposed to be focused is within the rendered tags range.
+		// If it is, make sure it is focused. If it is not - focus the tags list.
+		if (indices.includes(this.focusedTagIndex)) {
+			let tagsNodes = [...tagsList.querySelectorAll(".tag-selector-item")];
+			let tagToFocus = this.props.tags[this.focusedTagIndex];
+			let nodeToFocus = tagsNodes.find(node => node.textContent == tagToFocus.name);
+			if (nodeToFocus) {
+				nodeToFocus.focus();
+				if (this.keypressToHandleOnNextUpdate) {
+					this.shiftFocus(this.keypressToHandleOnNextUpdate);
+					this.keypressToHandleOnNextUpdate = null;
+				}
+			}
+		}
+		else {
+			tagsList.focus();
+		}
+	};
+
+	handleBlur = (event) => {
+		// If the focus leaves the tags list, clear the last focused tag index
+		let tagsList = document.querySelector('.tag-selector-list');
+		if (!tagsList.contains(event.relatedTarget)) {
+			this.focusedTagIndex = null;
+		}
+	};
+
+	handleKeyDown(e) {
+		if (!["ArrowRight", "ArrowLeft"].includes(e.key)) return;
+		// If the windowing kicks in, the node of the initially-focused tag may not
+		// exist, so first we may need to scroll to it.
+		if (!document.activeElement.classList.contains("tag-selector-item")) {
+			this.setState({ scrollToCell: this.focusedTagIndex });
```

Comment:
use `setState()` second param callback to continue handling of the event here instead of in `handleSectionRendered()`

Reply:
https://react.dev/reference/react/Component#setstate:~:text=optional%20callback%3A%20If%20specified%2C%20React%20will%20call%20the%20callback%20you%E2%80%99ve%20provided%20after%20the%20update%20is%20committed.

Reply:
That doesn't seem right. The `setState()` callback is called after rendering the state changes. Are you saying `Collection` double-renders? Once for scroll and once for rendering of tags?

If so, can we instead add a function like `waitForTagToRender()` that returns a promise which is resolved in `handleSectionRendered()`, such that all effects of the event handler can occur in the handler function/closure?

Reply:
> EDIT or does some delayed rendering of actual Collection nodes (tags)?

That is my impression. At least, `console.log` of all tag nodes in the callback of `setState` gives us the "stale" tags before scrolling, while logging them in `handleSectionRendered` yields current tags after scroll happen. I guess it makes sense, otherwise, why would `onSectionRendered` prop need to exist?

I pushed a change where we wait for the promise resolved in `handleSectionRendered` before shifting focus - it is indeed more readable that way